spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala

```diff
@@ -109,6 +112,64 @@ object Engine {
   private var physicalCoreNumber = -1
   private var nodeNum: Int = -1
 
+  @volatile
+  private var gatewayServer: py4j.GatewayServer = null
+
+
+  private[bigdl] def createJavaGateway(driverPort: Int): Unit = {
+    if (SparkUtils.isDriver) {
+      val file = new java.io.File(SparkFiles.getRootDirectory(), "gateway_port")
+      if (file.exists()) {
+        file.delete()
+      }
+      file.createNewFile()
+      val out = new PrintWriter(file)
+      out.print(driverPort)
+      out.flush()
+      out.close()
+      return
+    }
+    if (gatewayServer != null) return
+    this.synchronized {
+      if (gatewayServer != null) return
+      gatewayServer = new py4j.GatewayServer(null, 0)
+    }
+    val thread = new Thread(new Runnable() {
```

Comment:
what if another thread try to use this gateway while the gateway havn't inited

Reply:
Only one thread can reach this part of code.